table/block_based/block_based_table_builder.cc

```diff
@@ -1472,12 +1475,12 @@ Status BlockBasedTableBuilder::InsertBlockInCompressedCache(
 
 Status BlockBasedTableBuilder::InsertBlockInCacheHelper(
     const Slice& block_contents, const BlockHandle* handle,
-    BlockType block_type) {
+    BlockType block_type, bool is_top_level_filter_block) {
   Status s;
   if (block_type == BlockType::kData || block_type == BlockType::kIndex) {
     s = InsertBlockInCache<Block>(block_contents, handle, block_type);
   } else if (block_type == BlockType::kFilter) {
-    if (rep_->filter_builder->IsBlockBased()) {
+    if (rep_->filter_builder->IsBlockBased() || is_top_level_filter_block) {
```

Comment:
I think there's actually another issue here, namely the type for the old block-based filter seems also incorrect. The correct type would be `BlockContents`. (The old filter is block-based in the sense that it has separate Bloom filters for ranges of blocks but it's not `Block`-based :) )

I don't think it's a big deal since block-based filters are deprecated but would probably be nice to fix.

Reply:
We do have some lingering internal users IIRC

Reply:
Thanks @ltamasi for pointing it out. I miss understood your comment and thought its deprecated so don't need it anymore.